### PR TITLE
First draft of json-augmentation of labs

### PR DESCRIPTION
**Pull request by BrianLefler**
Would love feedback on this.  It adds two json map names with special treatment: lab_1side and lab_4side.

Any map which is added under those names will go into the mix to be randomly selected when creating labs.  They use the standard json generation, but add on proper handling for stairs and ice_lab settings.

They are mixed in with existing content in this way:
- lab_1side maps will be weighted alongside the 2 current 1-sided options (barracks & prisoner containment, each of which have a weight of 400).
- lab_4side maps will be used 20% of the time.

The maps that have been created use nested submaps extensively to populate a lot of smaller rooms, similar to the existing cross and tic-tac-toe boards.  But they aren't great and I'm not very good with json, so if anyone wants to make suggestions or changes, I'd love to see more.  I just wanted to get the ball rollling.

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
That is something we were all waiting for a long time - see #10958.

I will take a look asap.

**Comment by BrianLefler** [comment]
Cool, well the json runs alongside rather than replacing.  The thought would be as content came in, a greater portion of labs could become json-driven, until it becomes almost entirely json (leaving just stairs, cold, and ant/slime adjacency effects to be handled in code).  The 1sided rooms adjust naturally, the 4sided rooms require a tiny cpp change to make it more common.

I heavily used the nested json functionality to have variation--that feature's amazing and thank you I think to @Coolthulhu for doing it.

Right now it does not support rotation, seem to set traps, or allow non-square maps; all of which would make life a little easier if added and I'd like to dig into later.  I put rotation settings in nested json objects right now, which should just increase the observed variation whenever it gets implemented.

**Comment by BrianLefler** [comment]
Ok, I have cleaned up the adjacency json to ensure walls in all 4 directions (walls north&west, remove doors south&east).  It is a bit brutal in that it completely clobbers whatever's there, but it works.

I am a bit suspicious of the drop rates, which I suspect I've set a little high.  Any idea if it's possible / desirable to consolidate loot generation into a palette?  Then it could be reused through lab json and tweaked in one place to get the right level of loot.

File: data/json/mapgen/lab/lab.json [hunk 1]
```diff
@@ -0,0 +1,956 @@
...
+    }
+  },
+  {
+    "type": "mapgen",
+    "method": "json",
+    "nested_mapgen_id": "lab_north_wall",
+    "object": {
+      "mapgensize": [ 24, 24 ],
+      "place_terrain": [
+        { "ter": "t_concrete_wall", "x": 0, "y": 0 },
```
Comment (ZhilkinSerg) [inline data/json/mapgen/lab/lab.json]: I believe you can draw lines of terrain:

https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/MAPGEN.md#22-set

Comment (AMurkin) [reply data/json/mapgen/lab/lab.json]: Am i wrong that there is an error in the example of "line" in MAPGEN.md?
`Example: { "point": "terrain", "id": "t_lava", "x:" 5, "y": 5, "x2": 20, "y2": 20 }`

Comment (ZhilkinSerg) [reply data/json/mapgen/lab/lab.json]: Yes, docs can be outdated a bit.

**Comment by BrianLefler** [comment]
If I have time over the next week I can separate the json out into files (24x24 floorplans, 7x7/9x9 subrooms, palettes) and add more content.  I could also start migrating hardcoded content to json if people want.

Most useful for me is direction on what people want to see done here, or whether it should go in as a pull now and be added to later.

**Comment by ZhilkinSerg** [comment]
> Most useful for me is direction on what people want to see done here, or whether it should go in as a pull now and be added to later.

Great job so far! Will try to mergetest it during weekend and give some feedback.

**Comment by BrianLefler** [comment]
Added a few more various tweaks; the biggest is a new cross floor plan which fills in a quadrant with rocks when it borders a rock overmap tile, creating terrain like the attached screenshots.  Whenever a direction leads to rocks, that blueprint normally fills in that hallway, but also has a 25% chance of putting in a bionic vault (behind a high security console+door: If multiple vaults drop, there is only one console).  
screen shot 2018-07-02 at 8 43 36 am
screen shot 2018-07-02 at 8 43 10 am

**Comment by Night-Pryanik** [comment]
Consider adding a `[WIP]` prefix to PR's name if you are still working on it.

**Comment by Night-Pryanik** [comment]
Could you add a some sort of medical lab/operational theater with autodoc and the like, where scientists conducted experiments that resulted in creation of failed cyborgs? Or maybe just update pre-existing dissector rooms?

**Comment by BrianLefler** [comment]
> Consider adding a [WIP] prefix to PR's name if you are still working on it.

Ok.  I have been trying to keep it in a workable state that could be accepted at any point, but there's always more to do (particularly with json, since it is really easy to add to).

I will add an operating theater later today, making the operating lab a full 24x24 tile and not a little room so that it can have a weight assigned to it and be rare.  (There are only about a dozen 7x7 and dozen 9x9 rooms right now, so they often repeat, at least, they often repeat among the 20% of tiles which get the new json treatment).

The last big thing remaining I'd like to do is pull more of the hard-coded, non-random rooms out and convert them into json.  But that can happen as a separate PR as well.

**Comment by BrianLefler** [comment]
Ok, added.  It's a little unoriginal, but the operating theatre is basically a 180 rotation of the existing alien containment, with the option of a hard hack (security 6) to unlock to the autodoc room, or an easy hack (security 2) to unlock all doors and fight your way through four cyborgs.

**Comment by Night-Pryanik** [comment]
Cool, thank you so much.

**Comment by BrianLefler** [comment]
The hard-coded barracks and prisoner containment areas should now be in json, with the hardcoded versions removed.  This is the first time I've removed content instead of adding it, so it should be very obvious in testing.

Differences:
- containment rooms always have zombies, instead of a 50% chance of bodies (not sure how to do that in json).
- stairs are always outside security & floors inside security are blue.
- there is a 1-in-7 chance of the doors being smashed in by a hulk, which will 90% of the time be still there.
- slight differences in monster spawns (but it should be similar, json just has different levers).

Also, while I'm at it, in the barracks, shouldn't the magazine access control room have the doors reversed--so the locked door goes to the magazine and the unlocked door goes to the general barracks?  I have always smashed the window in and ignored the console, which seems like a mistake by whomever originally made the map.

**Comment by ZhilkinSerg** [comment]
Sorry, I didn't have time to look at this yet. Please inform when you think it will be ready to go and I will start testing. You could always add more changes in later PRs.

**Comment by BrianLefler** [comment]
Ok.  I am at a great stopping point.  The big hardcoded rooms for the lab have now all been converted to json.  It now does have a merge conflict, should I pull from master and resolve it or is that something you want to do?

Summary of big changes a player will notice:
- Semi-randomized json rooms, with cross, tic-tac-toe, etc layouts, and a dozen or so types of rooms (~20% of tiles).
- Barracks and prisoner containment now have blue floors and won't have stairs go into the locked areas.  1-in-7 chance a hulk will have broken the door.
- Barracks magazine console now needs to be hacked to get in, smashing the window won't help any more.
- The big rooms will now rotate after being converted to json (~20% of tiles).
- New bionic vault will appear rarely (will work out to about 1 in a 100 tiles).

**Comment by BrianLefler** [comment]
Ok, I have merged in master.  Should be good to review and I'm not planning any additional changes at this time.

**Comment by ZhilkinSerg** [comment]
Some furniture is spawned inside walls (look at those crates and tables):

image

**Comment by BrianLefler** [comment]
Yeah, I know!  I couldn't figure out a clean way to avoid this but it should not be visible to the user unless they take a jackhammer and start blasting through wall.  What's happening is that applying a submap with rows doesn't overwrite furniture.  So there's a room w/ furniture there in the normal cross-layout, then we decide to replace it with rocks & close the door (because it borders rocks).  The submap replaces the terrain w/ rocks, but not the furniture.

I think I can get rid of the furniture that by replacing the 9x9 submap with 81 individual place_terrain calls.  Will do that in a few days (as part of this PR or another).

**Comment by ZhilkinSerg** [comment]
> I think I can get rid of the furniture that by replacing the 9x9 submap with 81 individual place_terrain calls. Will do that in a few days (as part of this PR or another).

I guess that will be fine to fix in another PR.

**Pull request merged**